tsdb/shard.go

```diff
@@ -702,56 +702,64 @@ func (s *Shard) validateSeriesAndFields(points []models.Point, tracker StatsTrac
 			continue
 		}
 
-		name := p.Name()
-		mf := engine.MeasurementFields(name)
-
-		// Check with the field validator.
-		if err := ValidateFields(mf, p, s.options.Config.SkipFieldSizeValidation); err != nil {
-			switch err := err.(type) {
-			case PartialWriteError:
-				if reason == "" {
-					reason = err.Reason
+		cont, err := func(p models.Point, iter models.FieldIterator) (cont bool, err error) {
+			name := p.Name()
+			mf := engine.MeasurementFields(name)
+			mf.mu.Lock()
+			defer mf.mu.Unlock()
+			// Check with the field validator.
+			if err := ValidateFields(mf, p, s.options.Config.SkipFieldSizeValidation); err != nil {
```

Comment:
`ValidateFields` knows which fields are not currently in the measurement fields, and this is the only place in the code it is called from. We could avoid iterating over the fields again below looking for unknown fields if `ValidateFields` collected them for us.

Reply:
I agree. My question was how wide-ranging the changes should be. This PR is the minimal set of changes I could find that fixed the bug, but you are correct that there is much room for improvement in the code as it is in the product and in this PR.  Let's discuss how radical to get.

Reply:
I'd like to replace the `atomic.Value` storing the fields map with a `sync.Map` with a generic wrapper for type safety, for instance. Lots of locking goes away if we do that, but it's a big, scary change.